A32: gRPC xDS circuit breaking

voidzcy opened the pull request.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+resource configuration in Cluster Discovery Service (CDS) responses. Each
+circuit breaker [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds)
+defines settings for a RoutingPriority. gRPC only takes the first 
+_Thresholds_ with `priority` being [`DEFABUT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).
```
s/DEFABUT/DEFAULT/

voidzcy replied:
Fixed.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+limit.
+
+Envoy handles HTTP traffic with [connection pooling](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#connection-pooling),
+many circuit breaking limits are configured for connection/connection pool 
```
Please add the word "and" at the front of this line.

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+a single upstream endpoint address, those configurations do not apply to gRPC
+and will be ignored. See more in [Other configurations considered](#other-configurations-considered).
+
+As per upstream cluster configurations, circuit breakers will be implemented as
```
Suggest writing this paragraph as something like this:

With the introduction of [gRFC A28: xDS Traffic Splitting and Routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md), it became possible for there to be more than one CDS LB policy instance for a given cluster (e.g., if one route pointed to a given cluster and a different route split traffic between that cluster and other clusters).  However, after implementing the changes described in [gRPC A31: xDS RouteActions Support](https://github.com/grpc/proposal/pull/192), there will be only one CDS LB policy instance for a given cluster, which will make the CDS LB policy the logical place to implement circuit breaking.

voidzcy replied:
Rephrased.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+
+## Overview
+
+gRPC client receives circuit breaking configurations as part of the [Cluster](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#cluster) 
```
Please add "The" at the start of this line.

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+## Overview
+
+gRPC client receives circuit breaking configurations as part of the [Cluster](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#cluster) 
+resource configuration in Cluster Discovery Service (CDS) responses. Each
```
"resource received in the Cluster Discover Service (CDS) response"

voidzcy replied:
Rephrased.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+Envoy handles HTTP traffic with [connection pooling](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#connection-pooling),
+many circuit breaking limits are configured for connection/connection pool 
+management. Since each gRPC client creates at most one HTTP/2 connection to 
+a single upstream endpoint address, those configurations do not apply to gRPC
```
The terminology doesn't seem to be consistent throughout this document: I see the words "configurations", "limits", and "fields" being used interchangeably.  I think it will be easier for readers to understand if we pick one term and use it consistently throughout the document.

voidzcy replied:
Fixed to use "configurations". I proofread the whole doc, should have no obvious interchanged usages now.

markdroth replied:
Actually, I think the term "configurations" is confusing, because people don't tend to use that term to refer to individual configuration parameters; they tend to use that term to refer to the configuration as a whole.

How about using the term "parameter" instead?

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+### Maintaining thresholds
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
```
s/Make sure this threshold can be updated/This threshold will be updated/

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
+sure this threshold can be updated when receiving an update from the
+XdsClient cluster watch interface and the updated value can be dynamically 
```
Please add a comma before the "and".

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
+sure this threshold can be updated when receiving an update from the
+XdsClient cluster watch interface and the updated value can be dynamically 
```
s/can be/will be/

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+
+### Limiting outstanding requests
+
+Each CDS LB policy uses a counter to count the number of requests sent through
```
I think we should write this a little more generally, without dictating quite so much of the implementation details.  I suggest something like the following:

The CDS policy's picker will enforce the configured limit.  Each CDS LB policy will maintain a counter of the number of requests currently in flight to the associated cluster.  The counter will be incremented when a request is started on a subchannel and decremented when the request is complete.  When the picker is going to start a call, it will check whether the current number of in-flight requests is lower than the configured limit.  If so, the picker will do the pick as usual, returning a subchannel.  If not, it will fail the call with status UNAVAILABLE.

voidzcy replied:
Rephrased.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+Picker of downstream LB policies. If the threshold is reached, it returns
+a result that leads to the request being dropped with `UNAVAIABLE` status and
+never gets retried. Also this drop is recorded and aggregated to the 
+`total_dropped_requests` in the cluster-level stats to be reported in
```
Currently, the drops are recorded in the EDS policy, not the CDS policy.  Does this mean we need to start reporting drops from both places?  Or can we coaleasce this somehow?

voidzcy replied:
My bad, this wasn't clearly mentioned in the design doc. But yes, we would need to be able to record drops in CDS policy. This drop stats is somewhat different from drops in EDS policy. Drops made in EDS policy are aggregated to explicit categories specified by DropOverload policies received in the EDS response. Drops due to hitting circuit breaking limits are not classified by categories, they are just accumulated to the [`total_dropped_requests`](https://github.com/envoyproxy/envoy/blob/10fed474642a90760b2a73f866087ef12d67e542/api/envoy/config/endpoint/v3/load_report.proto#L154-L155) of the `ClusterStats`.

Slight changes for managing recorded load stats might be necessary. Upon receiving an CDS update, the CDS policy tends to create a child policy based on the received eds_service_name. At this time, we should create a load recorder for this cluster:eds_service as well. Since load recorders are managed by the XdsClient, it shouldn't be a problem to do so. I can write up a section in the design doc to clearly describe the full flow of load stats recording and how its APIs look like if you want.

markdroth replied:
If we do this in the CDS policy, then I think we should handle it by having the XdsClient provide a separate API for use by the CDS policy than the one it already provides for use by the EDS policy.  This is exactly analogous to the split we discussed as part of the xds policy refactoring to have the drop stats reported by the EDS policy and the actual request stats reported by the LRS policy.

Alternatively, we could handle this by moving the enforcement of circuit breaking down into the EDS policy instead, so that these drops can be reported at the same place as we report the other drops.  The only change we'd really need to make that work would be to add a field to the EDS policy's config indicating the circuit breaking `max_requests` limit.

voidzcy replied:
We should not move the enforcement of circuit breaking into the EDS policy, as an CDS policy may switch its child EDS policies when an updated `eds_service_name` is received. Requests sent through pickers outputted by all of this child EDS policies should be aggregated together and counted against the cluster-level threshold. 

We should just add a getter API on XdsClient for accessing the stats object. The stats object for each cluster:eds_service should have the same lifecycle as the corresponding EDS policy and have the CDS policy manages its lifecycle.

The workflow would be:
When the CDS policy receives an CDS update (includes information for eds_service_name, lrs_server_name), the stats object for this cluster:eds_service should be created along with instantiating the child EDS policy for this eds_service_name. The CDS policy manages to shutdown the old child EDS policy, it should destroy (un-reference) the stats object for it as well. Inside the EDS policy, it calls the getter API to access the stats object for recording drops if necessary. In this way, both CDS and EDS policies have access to the stats object. Moreover, having the CDS policy managing the lifecycle of the stats object seems to be more natural.

markdroth replied:
I don't think it makes sense to try to tightly couple the lifetime of this object between the two policies, especially since they are not reporting the same data anyway.  I think it would be better to just have a separate object for the drop stats reported by the CDS policy, just like we do for the drop stats reported by the EDS policy vs. the request stats reported by the LRS policy.

voidzcy replied:
The drop stats recorded in both policy are eventually put together for load reporting. The `total_dropped_requests` in `ClusterStats` is the sum of drops recorded by the CDS policy and all of the drops recorded by EDS policies. We are just calliing different APIs of the stats object to record drops in both policies, but their data will still be put together. So it makes sense to use the same object recording both kinds of drops. We are bookkeeping a group of stats object with each contains the stats for a single cluster:eds_service. Its scope, per cluster:eds_service, is same as that for the EDS policy. The meaning is when the CDS policy creates its child EDS policies, it also creates objects to record stats for each of them.

markdroth replied:
The fact that the stats are put together for load reporting is irrelevant.  That's also true of the stats reported by the EDS and LRS policies: they all go into the same LRS request.  But they get reported via separate objects.

I think it's better to have separate objects for reporting each type of data and then have the XdsClient merge the data when it constructs the load report.  That seems much cleaner to me than trying to tightly couple the lifetime of the object used by multiple LB policies.

voidzcy replied:
Alright, if you'd like a separate pair of add/remove APIs for stats object to be used by the CDS policy, that's fine. But I am more leaning towards leave this as an implementation detail and let implementors decide the approach based on how each language manages child policies in their LB policy implementations. As a design, we would just mention drops caused by circuit breakers should be recorded in `total_dropped_requests` for load reporting.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+- An increment should always have a corresponding decrement. A common mistake
+can be the counter is incremented in the `Pick` operation while the pick result
+is discarded later, which leads to the value being leaked.
+- The counter should only be incremented if the pick result indeed leads to a
```
This implies that the CDS picker should call the EDS picker first and then decide whether to discard the result.  That seems a little inefficient; we shouldn't need to go through the trouble of calling the EDS picker if we're just going to discard its result anyway.

I think that if the current number of in-flight calls is already at the limit, then we can immediately drop any new picks without calling the EDS policy.  There's no need to call the EDS policy if we're going to drop the call anyway.

voidzcy replied:
Yes you are right, but that's for the drop due to hitting the circuit breaker limit. The case I am talking about here is the EDS picker can also return a result that is a drop (a.k.a., contains no Subchannel, or even an error result). In that case, the counter should not be incremented as this will never leads to a RPC being sent. In Java, we probably do not have this concern as we manipulate the counter in stream tracer (a.k.a., a callback object that listens on stream events such as stream created/closed). But in C-core and Go, the counter is incremented in the Pick operation, so implementations should make sure the pick does lead to an RPC being sent before incrementing the counter

markdroth replied:
Ah, okay -- sorry, I misunderstood what this said when I first read it.

It's not clear to me why this needs to be pointed out, because if the CDS policy itself is not deciding to drop the request, it would certainly not update its counter.

In any case, as I mentioned elsewhere, I think we can omit some of the detailed implementation concerns from this doc, since they're mainly of concern to implementors.

voidzcy replied:
Deleted details that are not directly related to CDS policy and its picker.

markdroth replied:
I don't actually see anything deleted here.  There's still a lot of text related to implementation details.

I don't necessarily mind keeping these details here if you want, but how about moving them into their own subsection, with a heading like "Implementation Considerations"?

voidzcy replied:
Sure, I eliminated the whole paragraph. It is implementation specific, readers with only high-level background may not be able to get it. Those can (have been) be discussed in our own design doc for implementors.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+
+gRPC client receives circuit breaking configurations as part of the [Cluster](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#cluster) 
+resource configuration in Cluster Discovery Service (CDS) responses. Each
+circuit breaker [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds)
```
Suggest writing this as:

The `Cluster` resource encodes the circuit breaking configuration as a list of [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds) messages, where each message specifies the limits for a particular RoutingPriority.  gRPC will look only at the first entry in the list for priority [`DEFAULT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).

voidzcy replied:
Rephrased.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+defines settings for a RoutingPriority. gRPC only takes the first 
+_Thresholds_ with `priority` being [`DEFABUT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).
+
+As of the current gRPC design, only the `max_requests` field, which limits the 
```
Suggest writing this paragraph as:

Of the limits configured in the `Thresholds` message, gRPC will support only the `max_requests` limit, which sets the maximum number of requests that can be in flight to the cluster at any given time.  None of the other limits are applicable to gRPC; for details, see [Other configurations considered](#other-configurations-considered).

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+
+## Other configurations considered
+
+There are other configurations in the xDS circuit breakers, most of them do not
```
s/most/but most/

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+## Other configurations considered
+
+There are other configurations in the xDS circuit breakers, most of them do not
+apply to gRPC's use case and they will be ignored in the implementation. The
```
Please add a comma before the "and".

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,157 @@
+- `max_pending_requests`: the maximum number of requests that will be queued
+while waiting for a connection to be established.
+    - gRPC's _wait-for-ready_ semantics intend to avoid failing RPCs 
+    prematurely before a connection becomes available, requests are intended to
```
s/requests/and requests/

voidzcy replied:
Done.

voidzcy commented:
Thanks for pointing out so many wording issues. I've proofread the doc and rephrased inappropriate sentences.

Btw, I am also thinking about we might probably want to support the `max_pending_requests` circuit breaking configuration. The reason not to support it seems to be not strong enough as calls will be buffered even for non-wait-for-ready calls (wait-for-ready bypasses subchannel errors while an unestablished connection results calls to be buffered by any means). I have talked about the implementation for supporting `max_pending_requests` in go/grpc-xds-circuit-breaking-v2. Basically the CDS LB policy maintain another counter for buffered calls. The picker outputted by the CDS policy will check if the delegated downstream picker buffers the call. Increment the counter if it does. Stop buffering and drop calls when the counter reaches the threshold. This counter is decremented when the counter for outstanding requests is incremented.

If you also agree that we should support `max_pending_requests`, I can update the design doc and this gRFC to include that.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -66,27 +65,27 @@ thresholds to very high values, for example, to
+number of in-flight requests is lower than the configured limit. If so, the 
+picker will do the pick as usual, returning a subchannel. If not, it will 
+fail the call with status UNAVAILABLE. Such failed calls will not be retried,
+while they will be recorded to the total dropped requests counts and reported
```
s/while/but/

voidzcy replied:
Done.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -66,27 +65,27 @@ thresholds to very high values, for example, to
+number of in-flight requests is lower than the configured limit. If so, the 
+picker will do the pick as usual, returning a subchannel. If not, it will 
+fail the call with status UNAVAILABLE. Such failed calls will not be retried,
+while they will be recorded to the total dropped requests counts and reported
```
s/total dropped requests counts/`total_dropped_requests` count/

voidzcy replied:
Done.

voidzcy commented:
I rewrote some paragraphs of the doc (e.g., reasoning for not to support `max_pending_requests`, etc). Also, changed to use the word "parameter(s)" when referring to specific field of the circuit breakers. Thanks for the dedicated review!

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -42,19 +42,20 @@ make the CDS LB policy the logical place to implement circuit breakers.
-The `Cluster` resource encodes the circuit breaking configuration as a list of
+The `Cluster` resource encodes the circuit breaking parameters in a list of
 [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds)
 messages, where each message specifies the limits for a particular 
```
s/limits/parameters/

voidzcy replied:
Changed.

voidzcy replied:
Changed.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,137 @@
+
+## Detailed design
+
+### Maintaining thresholds
```
I suggest removing this sub-header and the "Limiting outstanding requests" one.  Both of these paragraphs can be together under the "Detailed design" header.

voidzcy replied:
Sure, cleaned up.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,137 @@
+
+### Maintaining thresholds
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
```
s/threshold/configured limit/

voidzcy replied:
Changed.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,137 @@
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. This
+threshold will be updated when receiving an update from the XdsClient cluster 
```
s/threshold/limit/

voidzcy replied:
Fixed, along with a couple of other "threshold" and changed them to "limit" for consistency.

voidzcy commented:
I updated the gRFC to reflect that circuit breakers will be implemented in EDS policy. PTAL.

htuch commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,147 @@
+Envoy handles HTTP traffic with [connection pooling](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#connection-pooling),
+and many circuit breaking parameters are configured for
+connection/connection pool management. Since each gRPC client creates at most
+one HTTP/2 connection to a single upstream endpoint address, those
```
Naive question; I had always thought that gRPC reserves the right to back a single "virtual" channel with multiple connections. Do you want to provide future proofing by reserving the right to add these multi-connection parameters in the future?

voidzcy replied:
Just from the design of [`Subchannel`](https://github.com/grpc/grpc-java/blob/e19bdf33b1d647c282169bc32a98362d53dd340b/api/src/main/java/io/grpc/LoadBalancer.java#L1201-L1217), it is intended to be a single connection at most. Although Java and Go does not have Subchannel sharing mechanism (C-core does have), we do have considerations trying to avoid cases that could potentially create more than one Subchannels to the same endpoint (e.g., [the RouteAction support design](https://github.com/grpc/proposal/pull/192) is trying to avoid having multiple CDS/EDS policies for the same cluster, and therefore a single leaf policy that creates actual connections to endpoints in the cluster). Based on that, I think it is safe to say we will not have more than one connections per endpoint. 

> Backing a single "virtual" channel (aka, Subchannel) with multiple connections.

That may be too far to be mentioned, unless the abstraction for connection/transport layer in gRPC is redesigned.

htuch replied:
Ack. It seems to be an implementation limitation but not some sort of in-principle limit.

markdroth replied:
I don't think anything in this design stops us from adding support for these parameters in the future if/when it becomes necessary.

For example, it's possible that in the future, we might add some new type of LB policy that would allow multiple connections to address the MAX_CONCURRENT_STREAMS limit from the server.  If/when we do that, we can reevaluate whether supporting this kind of circuit breaker makes sense.

htuch commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,147 @@
+application of the `max_requests` circuit breaking parameter in gRPC.
+
+To match Envoy's implementation, circuit breaking parameters will always be 
+enforced with certain defaults, even if the gRPC client receives no applicable 
```
Presumably certain defaults matches those of Envoy...

htuch commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,147 @@
+- `max_connection_pools`: gRPC does not use connection pooling.
+- `connect_timeout` (in `Cluster` configuration): hard to implement in gRPC's 
+architecture, will reconsider after getting usage feedback.
+- `max_requests_per_connection` (in `Cluster` configuration): lifetime limit 
```
I think there is general value in this one, since it allows you to put in place anti-hotspotting mitigations, i.e. forcing new connections after a certain amount of traffic. But, this can also be done with connection lifetime timeouts.

voidzcy replied:
> forcing new connections after a certain amount of traffic

Since we will create at most one connection per endpoint, this behavior would be shutting down the connection and re-establish. This doesn't seem to do anything helpful. Also we are not able to have a timely connection shutdown and re-establishment, we still need to let ongoing RPCs continue and finish. So during the period after the threshold is hit, the effective behavior would be the same as this Subchannel has been shut down. And it will be recreated again. The effective observation is the endpoint is _temporarily_ unreachable by new RPCs, and how long it will be unreachable depends on when ongoing RPCs sending to it finishes. This sounds to be a very weird thing.

htuch replied:
I agree that if you have this restriction of maximum of one connection, it's hard to sensibly implement this policy without blocking the channel during drain.

htuch commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,147 @@
+while waiting for a connection to be established.
+    - As a proxy, Envoy drops queued requests from downstream entities to avoid
+    unbounded congestion and memory usage of the process. But as part of the
+    application, it makes little sense for gRPC to drop its own requests.
```
Arguably an application that is doing async send would benefit from a cap on otherwise unbounded queueing?

voidzcy replied:
@markdroth and I had some discussion about the buffering limit in https://github.com/grpc/proposal/pull/200#pullrequestreview-452627462. I am also leaning towards supporting it, but as mentioned in that comment, it turns out to be more complicated and there are cases we need to consider when putting together with other existing gRPC features.

The reasoning here for not supporting it doesn't seem to be very convincing. Maybe I should say not to be supported in the immediate first round, as it is complicated, and we will be trying to support more complicated ones progressively.

htuch replied:
Yeah, I think having some distinction between the features that will _never_ be implemented and those that _might_ be implemented later would be great.

markdroth replied:
It wouldn't hurt to expand on the descriptions for these items a bit.  For example, for this particular item, instead of just saying "it makes little sense for gRPC", let's explain that it's complicated to implement and that it's not clear that the benefit is enough to justify the cost at this point.

voidzcy replied:
Updated. I also reordered items to put together things not applicable and things with implementation difficulties.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -63,14 +55,28 @@ limits to very high values, for example, to
+in the LB policy tree. Since each CDS LB policy creates exactly one child EDS
+LB policy for endpoint discovery, both of them are part of the load balancing
+logic for requests sent to the specific cluster. Circuit breakers can be 
+implemented in either policies. In order to easily record and report requests 
```
s/policies/policy/

voidzcy replied:
Done. Thanks!

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -80,13 +86,21 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
 outstanding requests.
-- In CDS LB policy, wrap the `SubchannelPicker` propagated from its child 
-EDS policy with the logic of checking the counter and make pick decision.
+- Wrap the `SubchannelPicker` propagated from each of EDS policy's child LB 
```
The EDS policy is already wrapping the pickers from its children to handle the drops requested in EDS, so no new wrapping is needed.  We just need to modify the existing wrapper to also enforce the new max_requests field.

voidzcy replied:
The actual implementation is still flexible. One could choose to wrap another layer for this feature or could choose to add the logic to the existing wrapper. It is just for pointing out where to add the logic.

I updated the description a bit to make this more neutral.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -80,13 +86,21 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
 
 ### Java
-- Maintain a counter in the CDS LB policy for counting the number of 
+- Add a `max_requests` field in the EDS LB config. The value will be taken
```
Please specify the exact protobuf changes for the EDS LB policy config.

voidzcy replied:
Sure. The proto change to the EDS LB policy config is general across languages, so I put it to the detailed design section.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -63,14 +55,28 @@ limits to very high values, for example, to
+in the LB policy tree. Since each CDS LB policy creates exactly one child EDS
+LB policy for endpoint discovery, both of them are part of the load balancing
+logic for requests sent to the specific cluster. Circuit breakers can be 
+implemented in either policies. In order to easily record and report requests 
```
We should probably explicitly say that the reason for doing this in the EDS policy is that that's where we're already handling the drops requested by EDS, so it makes sense to handle these drops in the same place.

voidzcy replied:
Updated.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -80,13 +86,21 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
 but they will be recorded to the `total_dropped_requests` counts in 
 cluster-level load reports and reported to the load reporting server.
 
+Note an EDS LB policy may be responsible for handling `service_name` switch
```
I think this may be difficult to implement, and I'm not sure it's actually necessary.  If the EDS service name is changing, it's also likely that the set of backends is changing, in which case it would be fine for the active requests to be counted separately for each set of backends.

I think we should just remove this.  If it actually becomes a problem (which seems very unlikely), we can address it later.

voidzcy replied:
Isn't our current EDS policy implementation already handling the EDS service name change? The stats are recorded separately for each set of backends (remember each `ClusterStats` are for stats per {cluster, eds_service}). But requests sent both EDS services should be aggregated together for circuit breaking limits, as logically they are requests sent to the same cluster. 

There is not really anything special to be added for the implementation, just highlighting that in-flight requests should be aggregated to the same counter for enforcing `max_request` while drops are recorded separately if they sent to different EDS services.

markdroth replied:
Recording the drop stats is not difficult to implement at all; as you said, we're already doing that for the EDS-based drop stats.  However, having a common counter for enforcing the max-requests limit may be tricky to implement.

In C-core, we completely swap out the EDS policy instance when the EDS service name changes, and we use fairly generic logic for switching between policies.  To have a common counter between them, we would need to have special logic specific to the EDS policy to reach into the old policy instance to get the counter and then pass it to the new policy instance that we're instantiating.  This is doable, but it does add complexity.

Also, there are some semantic questions here.  What happens if a single CDS update both changes the EDS service name and updates the max-requests limit at the same time?  In that case, should we apply the new limit only to the new EDS policy instance, not the old one, or should it immediately take effect for both?

And as I said above, it's not clear to me that it's really beneficial to share the counter between the two policies.  If we're changing the EDS policy name, we're probably also changing the set of backends, so applying different counters to the two instances seems fine anyway.

@htuch I'm curious as to what Envoy does in this case.  Does envoy track the max-requests circuit breaker per cluster or per EDS service name?  I wouldn't be surprised if it tracks it per cluster, given that the circuit breaker config is part of the CDS response and not the EDS response.  If so, how does Envoy handle the edge cases I mentioned above?

voidzcy replied:
> In C-core, we completely swap out the EDS policy instance when the EDS service name changes, and we use fairly generic logic for switching between policies.

Oh I see. We are having somewhat different implementations so that our definition for EDS policy is actually different. Here EDS policy, specifically I am really referring to the class that implements the `LoadBalancer` interface. In C-core a EDS policy's scope is bounded to a {cluster:eds_service} and switching eds_service happens outside the EDS policy (it's actually inside its CDS policy). However, in Java a EDS policy is somewhat "an extension of the CDS policy", as its scope is bounded to a cluster (same as the CDS policy) and internally it maintains a switchable subcomponent that takes care of the EDS service name switch. So our "EDS policy" is really just a wrapper boxing the real EDS policy and  instantiated by its parent CDS policy, it is not doing any load balancing work except receiving the EDS config. After that, the received config is delegated to the subcomponent.

Such an implementation is actually quite awkward and unclean. For example the wrapper class does nothing and has no real semantic meaning. It contains the work delegated by the CDS policy (aka, maintain two child real EDS policy instances during EDS service name switch). Also, this only happens on CDS and EDS polices (I believe part of the reason may be they are the two that evolved from the very early version of xDS design), each layer of the downstream policies below EDS policy maintains its child policies right inside its own implementation. I believe this implementation really needs to be refactored. After that, Java's definition of "EDS policy" should be clearer. 
/cc @dapengzhang0 who implemented the existing CDS/EDS policies.

> What happens if a single CDS update both changes the EDS service name and updates the max-requests limit at the same time? In that case, should we apply the new limit only to the new EDS policy instance, not the old one, or should it immediately take effect for both?

While waiting for @htuch's answer, my understanding is that the max-requests limit should be enforcing requests sent to hosts in a (virtual) cluster. If a cluster is changing its backing service (aka, EDS service), unfinished RPCs to the old service are still within the cluster. So they should definitely contribute to the number of in-flight requests to the cluster. Since we are doing graceful switch for EDS service change, the switching process is really first expand the cluster to two services and then drop the old one. That's why I think we should use a common counter for EDS policies having the same parent CDS policy. This also implies that when a single CDS update changes both the EDS service name and the max-requests limit, it should immediately take effect for both services.

In terms of having a common counter for enforcing the max-requests limit for EDS policies with the same parent CDS policy. It is tricky because we don't have a convenient way of passing arguments from a parent policy to its child policy. We use the Update method in `LoadBalancer` interface to do so (though this is another awkward thing in the design). It is unfortunate. We would need a slight change for existing description of implementing circuit breaking: we should maintain the counter in the CDS policy while passing it as an address attribute to its child EDS policy instances, so requests sent through EDS policies with the same parent CDS policy are aggregated to the same counter.

htuch replied:
Envoy tracks per-cluster. EDS service name is really just a convenience for allowing clusters to modify the name they use for EDS lookup. These aren't really first class concepts inside Envoy beyond this convenient realiasing.

htuch replied:
It's worth pointing out that circuit breakers are not an EDS concept; they apply to any cluster type that Envoy supports.

markdroth replied:
> It's worth pointing out that circuit breakers are not an EDS concept; they apply to any cluster type that Envoy supports.

That's a really good point.  We do plan to support at least LOGICAL_DNS and aggregate clusters in the future, and we might want to add other cluster types as well.  And in those cases, we would not be using the EDS LB policy.

Harvey, do circuit breakers apply to aggregate clusters as well?  How does that work if the underlying clusters also have their own circuit breakers defined?

voidzcy replied:
We want requests sent through the same CDS LB policy to be aggregated together and the total number is limited by the cluster-level circuit breakers. But on the other hand, among those requests dropped by circuit breakers should be recorded and reported separately for each {cluster, eds_service}, which is the scope of an EDS LB policy. Ideally we would implement this max-requests limit in the single Picker in CDS LB policy. But because we have to record drops separately for each eds_service, we need to split this work to each EDS LB policy's Picker while still aggregating to the same per-cluster counter. The aggregation is effectively equivalent as it is implemented in CDS LB policy's Picker, and conceptually it is tracking requests sent to the cluster. So I think the solution would be maintain the counter in the CDS LB policy and pass it down to the child EDS LB policies for enforcing the limit. With this approach, each EDS policy knows which requests are dropped by the circuit breaker while requests are still tracked per-cluster. @markdroth Does this sound good?

markdroth replied:
I want to fully understand how Envoy does this before we decide how to structure our implementation.  I don't want to go down a path that results in dramatically different semantics from Envoy.

> Envoy tracks per-cluster. EDS service name is really just a convenience for allowing clusters to modify the name they use for EDS lookup. These aren't really first class concepts inside Envoy beyond this convenient realiasing.

Is that also true for load reporting?  How does load reporting work for the case where the client receives a CDS update that changes the EDS service name?  Note that each load report in an LRS request contains both the cluster name and the EDS service name, which implies that both fields are used as the "key" for the corresponding load report data.  So I'd like to understand how Envoy splits up the load reporting information when the EDS service name changes.

For example, consider the following timeline (assumes 10s load reporting interval):

- t=0: Envoy sends an LRS request including a load report for {cluster_name=C, EDS_service_name=E1}.
- t=0 through t=3: Envoy sends 10 requests to the cluster and drops 2 requests.
- t=3: Envoy receives a CDS update changing the EDS service name from E1 to E2.  Envoy requests E2 via EDS.  (Presumably it continues to use the backends from E1 until it gets new backends from E2?)
- t=4: Envoy has now gotten the EDS resource E2 and connected to those backends, so it now switches over and stops using the original backends from E1.
- t=4 through t=10: Envoy sends 15 requests to the cluster and drops 4 requests.
- t=10: Envoy sends a new LRS request.

The question is, what is included in the LRS request sent at time t=10?  Note that there are two sets of stats here: those from before the EDS service name changed and those after the EDS service name changed.  If Envoy just sends one load report for the cluster as a whole and reports the most recent EDS service name, then it will send both sets of stats as a single entry, like this:

- {cluster_name=C, EDS_service_name=E2}: requests=25, drops=6

But if it records each set of stats under its own EDS service name, then it would send this:

- {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2
- {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

gRPC does the latter.  This seems more accurate, because if E1 and E2 represent disjoint sets of endpoints, then it does not really make sense to lump their stats together.  It also seems to violate the LRS semantics of using both the CDS and EDS names as the key.

By that same logic, it seems weird to track the max-requests counter at the CDS level, because the resulting stats would need to be tracked and reported separately for the two EDS service names.  This also implies that if we get a CDS update that both changes the max-requests limit and the EDS service name at the same time, the new limit should apply only to the new EDS service name.  (Although admittedly, that last part would be much clearer if the max-requests limit was set in the EDS response instead of the CDS response.)

I guess what I'm pointing out here is that it seems very inconsistent to say that the EDS resource name is not a first-class concept but that it is used as part of the key for the LRS data.  I think we need to clarify and document how this is supposed to work.

voidzcy replied:
Thanks @markdroth for posting out such a detailed example, this accurately questions what has been bothering me for a long time and I think it is very valuable for us to completely understand the whole logic in Envoy. @htuch Could you please help us figure this out? Thanks.

htuch replied:
For aggregate clusters, I believe we have just a single circuit breaker. There are no tests for this, I've opened up https://github.com/envoyproxy/envoy/issues/12739. I think it would also make sense to add a PR to clarify in the API.

htuch replied:
The EDS service name and LRS interaction is limited to https://github.com/envoyproxy/envoy/blob/cbdd01c22b84ac5333f95f030821dfcf6db49734/source/common/upstream/load_stats_reporter.cc#L77. 

When you change EDS service name, you're basically replacing the cluster. At that point, you are resetting all the stats.

voidzcy replied:
@htuch Thanks for your quick response!

> When you change EDS service name, you're basically replacing the cluster.

This makes sense to me. But I believe all the subtleties come from dealing with streaming RPCs and graceful switches for EDS service name change. When these two are involved, it's hard to make a clean replacement. 

> At that point, you are resetting all the stats.

Does this mean some stats are just thrown away when the transition happens? The situation is especially hard to make a clear cut when there are streaming RPCs ongoing to hosts in the old EDS service.

How does Envoy deal with the transition? For example, there are requests in-flight to hosts in the old EDS service and they it receives an CDS response to switch to another EDS service. Does it let the in-flight requests finish? What if the requests are streaming and never finish? Before connections to endpoints in the new EDS service are established, are new requests just get buffered instead of being sent to endpoints in the old EDS service? 

I think this would be very important to us, as our current implementation is trying to handle switchovers as mild as possible (e.g., trying to use the old EDS service before the new one becomes ready, trying to record/report every single load, etc). The effort trying to gracefully deal with the very detailed aspects of transitions makes our design somewhat diverge from what Envoy does, and some usages may not be intended in Envoy's design.

htuch replied:
I don't understand what "graceful switches for EDS service name change" means here. When a cluster is replaced, the new cluster first warms, and then when it completes warming, replaces the existing cluster. The older cluster connections are then drained. @snowp can most likely add some additional detail here (or correct any inaccuracy).

snowp replied:
The new cluster definitely replaces the old one completely after the new one is warmed, I'm not sure exactly what happens with the old one. It's possible that we just tear down the old connections without warning. 

In any case there is no handing over of connections etc., the new and old cluster are independent.

voidzcy replied:
> I don't understand what "graceful switches for EDS service name change" means here.

It's about when receiving an CDS response that contains a different EDS service name inside. Graceful switch means we will keep sending requests to hosts resolved by the old EDS service name until connection to hosts resolved by the new one is established. So this would be the same as what you are saying about warming the new cluster.

The problem is, "the older cluster connections are then drained", does it just forcefully terminate requests even if they are streaming? And I think this is something that is unclear in our design, although it is a very detailed thing. The time a streaming RPC will finish is undefined, if we let it keep going, we will wait for an unbounded amount of time to cut off the old cluster. That would keep the stats object for the old cluster for an unbounded amount of time as well. So does Envoy just throws away the unreported stats for the old cluster when it is cut off?

htuch replied:
AFAICT, Envoy transitions the connections to hosts in the respective connection pools on the no longer active cluster to drained state, but I don't see where GOAWAY is sent or any timeouts in the connection pool draining. This would imply they no longer contribute to LRS stats (or cluster stats), potentially indefinitely. @snowp @mattklein123 does this seem right?

mattklein123 replied:
The client doesn't send GOAWAY to upstream since we don't support push currently and it wouldn't do anything. Basically we just wait until there are no active streams per other mechanisms (they finish, timeout, etc.)

htuch replied:
In theory we could have a long running stream, e.g. several hours, where there is an observability gap as a result of this if I read ^^ right.

mattklein123 replied:
Yes, that case could be an issue. Presumably there is some idle timeout still in play in which it would drain eventually, but if we want active draining of streams a a result of the update I think that functionality would need to be added.

voidzcy replied:
> The question is, what is included in the LRS request sent at time t=10? Note that there are two sets of stats here: those from before the EDS service name changed and those after the EDS service name changed. If Envoy just sends one load report for the cluster as a whole and reports the most recent EDS service name, then it will send both sets of stats as a single entry, like this:
> 
> {cluster_name=C, EDS_service_name=E2}: requests=25, drops=6
> But if it records each set of stats under its own EDS service name, then it would send this:
> 
> {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2
> {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

@markdroth Based on what has been discussed above, how exactly stats are recorded/reported for requests sending to the old EDS service during the transition seems to be a bit vague, even in Enovy it might not have a definitive answer. I can imagine the subtleties. That amount of stats during the transition may not be that important and critical. But the answer to your question above (at time t=10, the stats being reported) seems to be either one of the following two:

- {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

or 

- {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2 
   {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

The former is we do not care the stats remain unreported for the old EDS service and the latter is we try to report every single load. Note for the latter report, the LRS server will combine IMO, the former is cleaner and easier to implement as the time a streaming RPC ends is unbounded.

Based on what is mentioned in https://github.com/grpc/proposal/pull/200#discussion_r473400202:
> When you change EDS service name, you're basically replacing the cluster (which is also true for other field mutations in the cluster). At that point, you are resetting all the stats.

Then it makes more sense that we use separate max-request counters for EDS policy (or reset the counter when changing to use a different EDS service name).

htuch replied:
I think we should fix Envoy's behavior to report both active and draining clusters, but this is more complicated than just thinking about EDS service name. There are other cluster updates that are possible that will result in similar behavior, but with no EDS service name change. I think in those cases, active/draining should both combine to form a unified cluster stats. I've filed a bug for this at https://github.com/envoyproxy/envoy/issues/12823.

voidzcy replied:
@htcuh Alright, that would affect at most how dropped requests will be reported. If Enovy choose to report stats via a unified cluster stats, then we need to change our existing behavior as well.

Back to our original question, does Envoy reset circuit breakers when receiving an CDS response that changes the EDS service name? This is more crucial for us about how the max-request enforcement should be applied.

htuch replied:
I'm pretty sure it does, since the circuit breaker resource tracking lives in the cluster, and EDS service name change implies a new cluster.

voidzcy replied:
Great, then that is clear enough to us for our existing issue. 

@markdroth You are right, we would need separate counters for each EDS LB to keep track of its in-flight RPCs.

markdroth replied:
(Sorry for taking so long to get back to this thread -- I've been really swamped with other work for the last couple of weeks.)

If I understand all of this discussion correctly, here's where things stand:

Based on the description in envoyproxy/envoy#12823, it seems like the LRS behavior that Envoy wants is what gRPC already has, which is to actually use both the cluster name and EDS service name as the key for the LRS data.  When the EDS service name changes, any activity for the old EDS service (e.g., drops recorded before the EDS service name changed or long-running streams started before the change) will be reported under the old name, and any activity for the new EDS service name will be reported under the new name.  So we don't need any changes to this existing code in gRPC.

Since any change in the Cluster resource basically causes Envoy to create a new Cluster instance, warm it, and then swap it out for the old one, that does imply that there is a separate counter of active requests for each Cluster instance.  So if gRPC tracks these drops in the EDS policy, then each EDS policy instance can have its own counter, and the counter will implicitly reset whenever we swap out the EDS policy, which is what we want.  And as an added bonus, I think this is much simpler for us to implement than any other behavior.

The only behavior difference between gRPC and Envoy will be that if there is a CDS update that changes some field *other* than the EDS service name (e.g., if it changes the TLS certificate config), Envoy would reset this counter, whereas gRPC would not.  But that actually seems better -- in principle, we should not reset this counter if the thing that changed was not part of the key for reporting the data to LRS.  However, it is worth noting that this could result in possibly surprising behavior: if the current number of RPCs in flight is 105 and we receive a CDS update lowering the limit to 100, we will immediately start failing all RPCs until 6 of the currently in-flight RPCs are finished.  Envoy would actually reset the counter here, which means that even if none of the original 105 RPCs ever completed, it would allow 100 new RPCs to be started.  I think this is okay, but we should probably call this out in the doc.

@voidzcy I think we need the following changes:
- Please add a note about the behavior difference I described above.
- Please remove the paragraph starting on line 105.

I think that with those changes, the gRFC should reflect what we want here.  Thanks to everyone for the discussion!

voidzcy replied:
Thanks for looking into this.

> if the current number of RPCs in flight is 105 and we receive a CDS update lowering the limit to 100, we will immediately start failing all RPCs until 6 of the currently in-flight RPCs are finished. Envoy would actually reset the counter here, which means that even if none of the original 105 RPCs ever completed, it would allow 100 new RPCs to be started.

I think some concepts are distorted in this description. Since Enovy _replaces the cluster_ when receiving a lowered circuit breaker limit, logically there are _old_ and _new_ clusters. New RPCs are sent through the _new cluster_ immediately (as connections are pooled) and the _new cluster_ starts with new circuit breaker tracking states. Then definitely another 100 new RPCs are allowed. But in gRPC there is only one logical cluster during this whole process, so new RPCs just continue the existing cluster's circuit breaker tracking states. So the problem is not because of how circuit breakers are applied and implemented, but really is because of the difference in the concept of _logical clusters_.

voidzcy replied:
I added the description in terms of this behavior difference. I think we should focusing on describing the high-level design and concepts instead of trying to clarify every single piece of detail.

markdroth replied:
The way you described this looks good.  I think it's primarily important to point out the actual behavior difference with this example, but also explaining the architectural reasons why the behavior difference occurs is also very useful.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -116,9 +130,11 @@ TODO
-There is a group of other parameters in the xDS circuit breakers, but most of
-them do not apply to gRPC's use case, and they will be ignored in the 
-implementation. The following parameters have been considered:
+There is a group of other parameters in the xDS circuit breakers, but they
```
I would phrase this a little differently.  I think the two categories are:

1. The parameters are fundamentally not applicable to gRPC in its current form.  It's possible that gRPC could evolve to make them relevant, in which case we would reevaluate them, but it's also possible that these will simply never be supported.
2. The parameters are applicable to gRPC, but the implementation is not trivial, and we don't want to invest the resources to implement them until/unless we know that they will be useful enough to make that investment worthwhile.  We will reevaluate these as we get feedback from users.

I suggest splitting this list of parameters into two lists, one for each of the categories above.

I think the following parameters are in category 1:
- `max_connections` and `max_requests_per_connection`: gRPC only ever creates a single connection to a given endpoint.  But in the future, we might provide a new LB policy to allow multiple connections to address MAX_CONCURRENT_STREAMS limits, at which point it might make sense to support this.
- `max_connection_pools`: gRPC does not currently have a concept of connection pools the way that Envoy does.  This will likely never apply to gRPC.

I think the following parameters are in category 2:
- `max_pending_requests`: Your existing description of this one is good.
- `connect_timeout`: In addition to what you've already written for this one, it also has particular challenges in C-core due to the way we share subchannels between channels.
- `max_retries` and `max_budget`: These will be addressed when we eventually add support for configuring retries via xDS.

voidzcy replied:
Updated to separate them into two categories.

voidzcy commented on A32-xds-circuit-breaking.md:
```diff
@@ -95,12 +111,10 @@ in separate stats message.
-- Wrap the `SubchannelPicker` propagated from each of EDS policy's child LB 
-policy with the logic of checking the counter and make pick decision.
+- Add the logic of enforcing `max_request` limit in wrapping the 
+`SubchannelPicker` propagated from each of EDS policy's child LB policy.
```
Fixed, thanks.

voidzcy commented on A32-xds-circuit-breaking.md:
```diff
@@ -64,16 +64,32 @@ there will be only one CDS LB policy instance for a given cluster at any time
+  repeated LoadBalancingConfig endpoint_picking_policy = 5;
+  
+  // Maximum number of outstanding requests can be made to the upstream cluster.
+  int32 max_requests = 6;
```
Sure, updated.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,162 @@
+  repeated LoadBalancingConfig endpoint_picking_policy = 5;
+  
+  // Maximum number of outstanding requests can be made to the upstream cluster.
+  int32 max_concurrent_requests = 6;
```
xDS uses `google.protobuf.UInt32Value` for the type of this field.  We should probably support the full unsigned integer range for compatibility.

We also need to document the default here, and we need to document how to disable this completely.  Envoy seems to disable it by setting it to the max value, which seems a little odd -- maybe we should use 0 for that instead?

voidzcy replied:
Envoy just uses a large number, so large that realistically not possible to achieve, to _effectively_ disable circuit breaking. This sounds legit and I don't think it's odd. One good thing for this is that implementations do not need to have special cases for the value. The value being 0? I guess a user will never set 0 for circuit breaking configurations if its usage is not specially noted, so a design doesn't need to focus on such extreme details and leave them as implementation flexibilities.

htuch replied:
I think you want to be consistent with Envoy here, control planes shouldn't have to think about two different values for disabling max concurrent requests.

markdroth replied:
I absolutely agree that gRPC should not use a different "disable" value in the xDS resource than Envoy does, because we don't want control planes to have to send different data to different clients.  My question here was about how we represent this internally (the config passed from the CDS LB policy to the EDS LB policy), which could be different from how it's represented in xDS.  But I do also see the argument that using the same representation here makes sense, for consistency.

That having been said, I guess it's also worth asking whether we should allow using a 0 value to disable the limit even in xDS.  Using the maximum allowed value for a uint32 seems sub-optimal for a few reasons:

- It's fairly cumbersome for humans to populate, because people have to figure out what value to use.  (I know I never remember what MAX_UINT32 is; I always have to look it up or calculate it.)
- Do we actually know for certain that there will never be MAX_UINT32 concurrent requests in a given cluster?  The cluster could have a really large number of endpoints, and on a busy proxy on a really beefy machine, there really isn't any predefined upper bound here.  It seems sort of arbitrary to impose a limit just due to not having another way to disable this.
- From a performance perspective, setting the limit really high isn't really the same thing as turning off tracking completely, because there's still some overhead involved.  Ideally there should be a way to say "don't even bother tracking this, because I don't want the overhead".  (I suppose an implementation could special-case MAX_UINT32 to provide that behavior, but that seems a little hacky too.)

For all of these reasons, I am wondering if we should change xDS to allow using 0 to disable these circuit breakers.  That seems safe enough, since I can't think of any legitimate use-cases where one would actually want to set the limit to 0, since that would basically disallow all requests.

Harvey, thoughts...?

htuch replied:
I guess 0 might be used today by folks who want to disable all connections, but there's more direct ways to do that. Technically this would be a breaking API change.

markdroth replied:
Yeah.  Okay, I guess let's not mess with this for now.  If we change it later, we can make that change in both Envoy and gRPC at that point.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -102,11 +103,15 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
+the latest configurations. The new cluster starts with new circuit breaker
+tracking states. However, in gRPC this only happens when the cluster's EDS
+service name changes. This could result in a surprising behavior compared to
+Enovy: if the current number of RPCs in flight is 105 and we receive an CDS
```
s/an CDS/a CDS/

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -102,11 +103,15 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
-should be aggregated to the same counter and the total is restricted by the
-configured limit. Even though the failed requests will be recorded and reported
-in separate stats message.
+Note in Envoy's definition, cluster resources are immutable and any change in
```
s/Note/Note that/
s/definition/implementation/

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -102,11 +103,15 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
-in separate stats message.
+Note in Envoy's definition, cluster resources are immutable and any change in
+the cluster resource implies replacing the old cluster with a new one with 
+the latest configurations. The new cluster starts with new circuit breaker
```
s/configurations/configuration/

markdroth reviewed (approved):
This looks great!

voidzcy commented:
Adding @ejona86 and @dfawley if there are any comments with the context. Will merge if no more concerns.

markdroth commented on A32-xds-circuit-breaking.md:
```diff
@@ -61,12 +61,12 @@ given cluster (e.g., if one route pointed to a given cluster and a different
-in EDS LB policy, it makes sense to handle circuit breaking drops in the same
-place.
+in the LB policy tree. Since the EDS LB policy is the counterpart for endpoint
+discovery of the cluster, both of CDS and EDS LB policies comprise the load
```
s/both of/both the/

ejona86 commented on A32-xds-circuit-breaking.md:
```diff
@@ -0,0 +1,167 @@
+Adopt Envoy's [circuit breaking](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking#circuit-breaking) 
+component that comes with the xDS protocol to
+fail network operations quickly and apply back pressure downstream as soon as
+possible. This allows gRPC clients to received distributed traffic limit 
```
s/received/receive/
